File: 99-use-cases/01-abcd-test/node/01-abcd-test.js
#!/usr/bin/env node
/** Serve the ABCD navigation count label grid navigator web UI on a local HTTP server. */

const http = require("http");
const fs = require("fs");
const path = require("path");
const { initClient, closeClient, evaluateCountLabel } = require("../abcd-eval.js");

// LaunchDarkly capability: String flag evaluation for multivariate A/B/C/D test
// See: https://launchdarkly.com/docs/sdk/features/flag-types

const PORT = 8080;
const ROOT = __dirname;

function sendJson(res, status, body) {
  const payload = JSON.stringify(body);
  res.writeHead(status, {
    "Content-Type": "application/json",
    "Content-Length": Buffer.byteLength(payload),
  });
  res.end(payload);
}

async function runServer() {
  await initClient();

  const server = http.createServer(async (req, res) => {
    const url = new URL(req.url, `http://127.0.0.1:${PORT}`);

    if (url.pathname === "/api/count-label") {
      const username = (url.searchParams.get("username") || "").trim();
      if (!username) {
        sendJson(res, 400, { error: "username query parameter is required" });
        return;
      }
      const countLabel = await evaluateCountLabel(username);
      sendJson(res, 200, { countLabel });
      return;
    }

    const urlPath = url.pathname === "/" ? "/index.html" : url.pathname;
    const filePath = path.join(ROOT, urlPath);

    if (!filePath.startsWith(ROOT)) {
      res.writeHead(403);
      res.end("Forbidden");
      return;
    }

    fs.readFile(filePath, (err, data) => {
      if (err) {
        res.writeHead(404);
        res.end("Not found");
        return;
      }
      const ext = path.extname(filePath);
      const type = ext === ".html" ? "text/html" : "text/plain";
      res.writeHead(200, { "Content-Type": type });
      res.end(data);
    });
  });

  server.listen(PORT, "127.0.0.1", () => {
    console.log(`Grid navigator (ABCD count label) running at http://127.0.0.1:${PORT}/`);
    console.log("Press Ctrl+C to stop.");
  });
}

const args = process.argv.slice(2);
if (args.length >= 2 && args[0] === "--evaluate-once") {
  (async () => {
    await initClient();
    try {
      console.log(await evaluateCountLabel(args[1]));
    } finally {
      await closeClient();
    }
    process.exit(0);
  })();
} else {
  runServer().catch((err) => {
    console.error(err);
    process.exit(1);
  });
}

process.on("SIGINT", async () => {
  await closeClient();
  process.exit(0);
});

File: 99-use-cases/01-abcd-test/python/01-abcd-test.py
#!/usr/bin/env python3
"""Serve the ABCD test grid navigator web UI on a local HTTP server."""

import json
import os
import sys
from http.server import HTTPServer, SimpleHTTPRequestHandler
from pathlib import Path
from urllib.parse import parse_qs, urlparse

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
from abcd_eval import close_client, evaluate_count_label, init_client  # noqa: E402

ROOT = Path(__file__).parent


class Handler(SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=str(ROOT), **kwargs)

    def do_GET(self) -> None:
        parsed = urlparse(self.path)
        if parsed.path == "/api/count-label":
            params = parse_qs(parsed.query)
            username = (params.get("username") or [""])[0].strip()
            if not username:
                self.send_error(400, "username query parameter is required")
                return
            body = json.dumps({"countLabel": evaluate_count_label(username)}).encode("utf-8")
            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.send_header("Content-Length", str(len(body)))
            self.end_headers()
            self.wfile.write(body)
            return
        super().do_GET()


def main() -> None:
    init_client()
    port = int(os.environ.get("PORT", "8080"))
    server = HTTPServer(("", port), Handler)
    print(f"Serving on http://localhost:{port}/", flush=True)
    try:
        server.serve_forever()
    finally:
        close_client()


if __name__ == "__main__":
    if len(sys.argv) >= 3 and sys.argv[1] == "--evaluate-once":
        init_client()
        try:
            print(evaluate_count_label(sys.argv[2]))
        finally:
            close_client()
        sys.exit(0)

    main()
